test-grapesy/Test/Sanity/Exception.hs

```diff
@@ -0,0 +1,196 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -Wno-orphans   #-}
+
+-- | Handling of exceptions occurring in an RPC on a shared connection.
+--
+-- These tests check for the behavior described in
+-- <https://github.com/well-typed/grapesy/issues/102>. In particular, there are
+-- two conditions that should hold when an exception occurs in the scope of a
+-- call (on either the server or client, e.g. inside either 'mkRpcHandler' or
+-- 'withRPC'):
+--
+-- 1. Other ongoing calls on that connection are not terminated, and
+-- 2. future calls are still possible.
+module Test.Sanity.Exception where
+
+import Control.Concurrent.Async
+import Control.Exception
+import Control.Monad
+import Data.Either
+import Data.IORef
+import Data.Text qualified as Text
+import Data.Word
+import Test.Tasty
+import Test.Tasty.HUnit
+
+import Network.GRPC.Client qualified as Client
+import Network.GRPC.Client.Binary qualified as Binary
+import Network.GRPC.Common
+import Network.GRPC.Server qualified as Server
+import Network.GRPC.Server.Binary qualified as Binary
+import Network.GRPC.Spec
+import Test.Driver.ClientServer
+import Test.Util.Exception
+
+tests :: TestTree
+tests = testGroup "Test.Sanity.Exception" [
+      testCase "client" test_clientException
+    , testCase "server" test_serverException
+
+    , testCase "earlyTerminationNoWait"  test_earlyTerminationNoWait
+    ]
+
+-- | Client makes many concurrent calls, throws an exception during one of them.
+test_clientException :: IO ()
+test_clientException = testClientServer $ ClientServerTest {
+      config = def
+    , client = simpleTestClient $ \conn -> do
+        -- Make 100 concurrent calls. 99 of them counting to 50, and one
+        -- more that throws an exception once it reaches 10.
+        let
+          predicate  = (> 50)
+          predicates =
+            replicate 99 predicate ++
+              [ \n ->
+                     (n > 10)
+                  && throw (DeliberateException $ SomeClientException 1)
+              ]
+
+        results <-
+          mapConcurrently
+            (   try @DeliberateException
+              . Client.withRPC conn def (Proxy @Trivial)
+              . countUntil
+            )
+            predicates
+
+        -- Only one of the calls failed
+        assertEqual "" (length $ lefts results) 1
+
+        -- All others terminated with results satisfying the predicate
+        assertBool "" (all predicate $ rights results)
+
+        -- New calls still succeed
+        assertBool "" . predicate
```

Comment:
Very nice, this is also a very pretty that verifies exactly what we wanted to verify :)